File: c/src/integration.c
/*
 * integration.c — CircleAI.Integration (C11 port of Contracts.cs).
 *
 * Record deep-copy / free primitives for the six integration records, shared by
 * the in-memory calendar / email / news / geo / home sub-modules. The interface
 * vtables themselves are plain function-pointer structs (declared in the header);
 * their instances are minted by the sub-modules. Pure C11 + libc. No pthreads.
 */

#include "circle_ai/integration.h"
#include "board_common.h"

/* ── nullable-string helper ─────────────────────────────────────────────── */

/* Copy a C# nullable string into (has,*dst). false only on OOM. */
static bool opt_copy(bool *has_dst, char **dst, bool has_src, const char *src) {
    if (has_src) {
        *dst = cab_strdup_empty(src);
        if (!*dst) { *has_dst = false; return false; }
        *has_dst = true;
    } else {
        *has_dst = false;
        *dst = NULL;
    }
    return true;
}

/* ── CalendarEvent ──────────────────────────────────────────────────────── */

void ca_int_calendar_event_free(ca_int_calendar_event_t *e) {
    if (!e) return;
    free(e->event_id);
    free(e->calendar_id);
    free(e->title);
    free(e->description);
    free(e->location);
    cab_strv_free(e->attendees, e->attendees_count);
    e->event_id = e->calendar_id = e->title = e->description = e->location = NULL;
    e->attendees = NULL;
    e->attendees_count = 0;
    e->has_description = e->has_location = e->is_all_day = false;
}
void ca_int_calendar_event_free_array(ca_int_calendar_event_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_int_calendar_event_free(&arr[i]);
    free(arr);
}

bool ca_int_calendar_event_copy(ca_int_calendar_event_t *dst,
                                const ca_int_calendar_event_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->event_id     = cab_strdup_empty(src->event_id);
    dst->calendar_id  = cab_strdup_empty(src->calendar_id);
    dst->title        = cab_strdup_empty(src->title);
    dst->start_utc_ms = src->start_utc_ms;
    dst->end_utc_ms   = src->end_utc_ms;
    dst->is_all_day   = src->is_all_day;
    bool ok = dst->event_id && dst->calendar_id && dst->title;
    ok = ok && opt_copy(&dst->has_description, &dst->description, src->has_description, src->description);
    ok = ok && opt_copy(&dst->has_location,    &dst->location,    src->has_location,    src->location);
    ok = ok && cab_strv_copy(&dst->attendees, src->attendees, src->attendees_count);
    if (ok) dst->attendees_count = src->attendees_count;
    if (!ok) { ca_int_calendar_event_free(dst); return false; }
    return true;
}

/* ── EmailMessage ───────────────────────────────────────────────────────── */

void ca_int_email_message_free(ca_int_email_message_t *m) {
    if (!m) return;
    free(m->message_id);
    free(m->from);
    cab_strv_free(m->to, m->to_count);
    free(m->subject);
    free(m->body_text);
    cab_strv_free(m->labels, m->labels_count);
    m->message_id = m->from = m->subject = m->body_text = NULL;
    m->to = m->labels = NULL;
    m->to_count = m->labels_count = 0;
    m->unread = false;
}
void ca_int_email_message_free_array(ca_int_email_message_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_int_email_message_free(&arr[i]);
    free(arr);
}

bool ca_int_email_message_copy(ca_int_email_message_t *dst,
                               const ca_int_email_message_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->message_id      = cab_strdup_empty(src->message_id);
    dst->from            = cab_strdup_empty(src->from);
    dst->subject         = cab_strdup_empty(src->subject);
    dst->body_text       = cab_strdup_empty(src->body_text);
    dst->received_utc_ms = src->received_utc_ms;
    dst->unread          = src->unread;
    bool ok = dst->message_id && dst->from && dst->subject && dst->body_text;
    ok = ok && cab_strv_copy(&dst->to, src->to, src->to_count);
    if (ok) dst->to_count = src->to_count;
    ok = ok && cab_strv_copy(&dst->labels, src->labels, src->labels_count);
    if (ok) dst->labels_count = src->labels_count;
    if (!ok) { ca_int_email_message_free(dst); return false; }
    return true;
}

/* ── NewsItem ───────────────────────────────────────────────────────────── */

void ca_int_news_item_free(ca_int_news_item_t *n) {
    if (!n) return;
    free(n->item_id);
    free(n->source_id);
    free(n->title);
    free(n->summary);
    free(n->url);
    cab_strv_free(n->tags, n->tags_count);
    n->item_id = n->source_id = n->title = n->summary = n->url = NULL;
    n->tags = NULL;
    n->tags_count = 0;
}
void ca_int_news_item_free_array(ca_int_news_item_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_int_news_item_free(&arr[i]);
    free(arr);
}

bool ca_int_news_item_copy(ca_int_news_item_t *dst, const ca_int_news_item_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->item_id          = cab_strdup_empty(src->item_id);
    dst->source_id        = cab_strdup_empty(src->source_id);
    dst->title            = cab_strdup_empty(src->title);
    dst->summary          = cab_strdup_empty(src->summary);
    dst->url              = cab_strdup_empty(src->url);
    dst->published_utc_ms = src->published_utc_ms;
    bool ok = dst->item_id && dst->source_id && dst->title && dst->summary && dst->url;
    ok = ok && cab_strv_copy(&dst->tags, src->tags, src->tags_count);
    if (ok) dst->tags_count = src->tags_count;
    if (!ok) { ca_int_news_item_free(dst); return false; }
    return true;
}

/* ── WeatherSample ──────────────────────────────────────────────────────── */

void ca_int_weather_sample_free(ca_int_weather_sample_t *w) {
    if (!w) return;
    free(w->condition);
    w->condition = NULL;
}
void ca_int_weather_sample_free_array(ca_int_weather_sample_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_int_weather_sample_free(&arr[i]);
    free(arr);
}

bool ca_int_weather_sample_copy(ca_int_weather_sample_t *dst,
                                const ca_int_weather_sample_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->at_utc_ms    = src->at_utc_ms;
    dst->temp_c       = src->temp_c;
    dst->feels_like_c = src->feels_like_c;
    dst->precip_mm    = src->precip_mm;
    dst->wind_kph     = src->wind_kph;
    dst->cloud_pct    = src->cloud_pct;
    dst->condition    = cab_strdup_empty(src->condition);
    if (!dst->condition) { ca_int_weather_sample_free(dst); return false; }
    return true;
}

/* ── RouteEstimate ──────────────────────────────────────────────────────── */

void ca_int_route_estimate_free(ca_int_route_estimate_t *r) {
    if (!r) return;
    free(r->polyline);
    r->polyline = NULL;
    r->polyline_count = 0;
}

bool ca_int_route_estimate_copy(ca_int_route_estimate_t *dst,
                                const ca_int_route_estimate_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->distance_km = src->distance_km;
    dst->duration_ms = src->duration_ms;
    if (src->polyline_count > 0) {
        dst->polyline = (ca_int_route_point_t *)malloc(
            src->polyline_count * sizeof(ca_int_route_point_t));
        if (!dst->polyline) return false;
        memcpy(dst->polyline, src->polyline,
               src->polyline_count * sizeof(ca_int_route_point_t));
        dst->polyline_count = src->polyline_count;
    }
    return true;
}

/* ── HaEntity ───────────────────────────────────────────────────────────── */

static void attr_pairs_free(ca_int_attr_pair_t *p, size_t n) {
    if (!p) return;
    for (size_t i = 0; i < n; ++i) {
        free(p[i].key);
        free(p[i].value);
    }
    free(p);
}

/* Deep-copy a pair array (each key/value empty-coalesced). false on OOM. */
static bool attr_pairs_copy(ca_int_attr_pair_t **out, const ca_int_attr_pair_t *src,
                            size_t n) {
    *out = NULL;
    if (n == 0) return true;
    ca_int_attr_pair_t *v = (ca_int_attr_pair_t *)calloc(n, sizeof(*v));
    if (!v) return false;
    for (size_t i = 0; i < n; ++i) {
        v[i].key   = cab_strdup_empty(src ? src[i].key : NULL);
        v[i].value = cab_strdup_empty(src ? src[i].value : NULL);
        if (!v[i].key || !v[i].value) { attr_pairs_free(v, i + 1); return false; }
    }
    *out = v;
    return true;
}

void ca_int_ha_entity_free(ca_int_ha_entity_t *e) {
    if (!e) return;
    free(e->entity_id);
    free(e->friendly_name);
    free(e->domain);
    free(e->state);
    attr_pairs_free(e->attributes, e->attributes_count);
    e->entity_id = e->friendly_name = e->domain = e->state = NULL;
    e->attributes = NULL;
    e->attributes_count = 0;
}
void ca_int_ha_entity_free_array(ca_int_ha_entity_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_int_ha_entity_free(&arr[i]);
    free(arr);
}

bool ca_int_ha_entity_copy(ca_int_ha_entity_t *dst, const ca_int_ha_entity_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->entity_id     = cab_strdup_empty(src->entity_id);
    dst->friendly_name = cab_strdup_empty(src->friendly_name);
    dst->domain        = cab_strdup_empty(src->domain);
    dst->state         = cab_strdup_empty(src->state);
    bool ok = dst->entity_id && dst->friendly_name && dst->domain && dst->state;
    ok = ok && attr_pairs_copy(&dst->attributes, src->attributes, src->attributes_count);
    if (ok) dst->attributes_count = src->attributes_count;
    if (!ok) { ca_int_ha_entity_free(dst); return false; }
    return true;
}

const char *ca_int_ha_entity_attr(const ca_int_ha_entity_t *e, const char *key) {
    if (!e || !key) return NULL;
    for (size_t i = 0; i < e->attributes_count; ++i)
        if (cab_ord_eq(e->attributes[i].key, key)) return e->attributes[i].value;
    return NULL;
}

File: rust/src/integration.rs
//! integration — CircleAI integration reference connectors.
//!
//! Full Rust port of `src/CircleAI.Integration/InMemoryIntegrationConnectors.cs`
//! together with the connector contracts they implement: deterministic,
//! dependency-free in-memory reference implementations of the integration
//! connector seams. These are the canonical offline/test doubles for calendar,
//! email, news, weather, routing, and home-automation, usable without any
//! external provider.
//!
//! The numeric connectors ([`InMemoryWeatherProvider`], [`InMemoryRoutingProvider`])
//! reproduce the C# formulas byte-for-byte, including .NET's banker's rounding
//! (`Math.Round`, round-half-to-even) via [`round_even`].
//!
//! Sync-only (the C# `ValueTask` methods collapse to plain returns, matching this
//! crate's other ports); `DateTimeOffset` → [`chrono::DateTime<Utc>`]; `TimeSpan`
//! → [`chrono::Duration`].
//!
//! Note: the briefing service in [`crate::companion`] carries its own reduced
//! connector seams tailored to that pipeline; these are the full-contract
//! reference connectors mirroring `CircleAI.Integration`.

use std::collections::HashMap;
use std::f64::consts::PI;
use std::sync::Mutex;

use chrono::{DateTime, Duration, TimeZone, Utc};

/// Rounds `value` to `digits` decimal places using round-half-to-even (banker's
/// rounding), matching .NET's default `Math.Round(double, int)`.
pub fn round_even(value: f64, digits: u32) -> f64 {
    let factor = 10f64.powi(digits as i32);
    let scaled = value * factor;
    let floor = scaled.floor();
    let diff = scaled - floor;
    let rounded = if (diff - 0.5).abs() < f64::EPSILON {
        // Exactly halfway: round to the even neighbour.
        if (floor as i64) % 2 == 0 {
            floor
        } else {
            floor + 1.0
        }
    } else {
        scaled.round()
    };
    rounded / factor
}

// ─────────────────────────────────────────────────────────────────────────────
// Calendar
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) A calendar event.
///
/// Mirrors the `CalendarEvent` shape consumed by `ICalendarConnector`
/// (`EventId`, `StartUtc`, `EndUtc`, `Title`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CalendarEvent {
    pub event_id: String,
    pub title: String,
    pub start_utc: DateTime<Utc>,
    pub end_utc: DateTime<Utc>,
}

impl CalendarEvent {
    /// Constructs a calendar event.
    pub fn new(
        event_id: impl Into<String>,
        title: impl Into<String>,
        start_utc: DateTime<Utc>,
        end_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            event_id: event_id.into(),
            title: title.into(),
            start_utc,
            end_utc,
        }
    }
}

/// (Integration) A calendar connector.
///
/// Mirrors `interface ICalendarConnector`.
pub trait ICalendarConnector {
    /// A stable provider identifier.
    fn provider_id(&self) -> String;
    /// Whether the connector is configured.
    fn is_configured(&self) -> bool;
    /// Events overlapping `[from_utc, to_utc)`, ordered by start.
    fn list_events(&self, from_utc: DateTime<Utc>, to_utc: DateTime<Utc>) -> Vec<CalendarEvent>;
    /// Creates (or overwrites) an event and returns it.
    fn create_event(&self, ev: CalendarEvent) -> CalendarEvent;
    /// Deletes an event by id (calendar id is ignored by the in-memory double).
    fn delete_event(&self, calendar_id: &str, event_id: &str);
}

/// (Integration) In-memory [`ICalendarConnector`]: events held in a map; listing
/// returns those overlapping the window, ordered by start.
///
/// Mirrors `sealed class InMemoryCalendarConnector`.
pub struct InMemoryCalendarConnector {
    events: Mutex<HashMap<String, CalendarEvent>>,
}

impl InMemoryCalendarConnector {
    /// Creates an empty connector.
    pub fn new() -> Self {
        Self {
            events: Mutex::new(HashMap::new()),
        }
    }
}

impl Default for InMemoryCalendarConnector {
    fn default() -> Self {
        Self::new()
    }
}

impl ICalendarConnector for InMemoryCalendarConnector {
    fn provider_id(&self) -> String {
        "in-memory".to_string()
    }

    fn is_configured(&self) -> bool {
        true
    }

    fn list_events(&self, from_utc: DateTime<Utc>, to_utc: DateTime<Utc>) -> Vec<CalendarEvent> {
        let mut hits: Vec<CalendarEvent> = self
            .events
            .lock()
            .unwrap()
            .values()
            .filter(|e| e.start_utc < to_utc && e.end_utc > from_utc)
            .cloned()
            .collect();
        hits.sort_by(|a, b| a.start_utc.cmp(&b.start_utc));
        hits
    }

    fn create_event(&self, ev: CalendarEvent) -> CalendarEvent {
        self.events
            .lock()
            .unwrap()
            .insert(ev.event_id.clone(), ev.clone());
        ev
    }

    fn delete_event(&self, _calendar_id: &str, event_id: &str) {
        self.events.lock().unwrap().remove(event_id);
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Email
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) An email message.
///
/// Mirrors the `EmailMessage` shape consumed by `IEmailConnector` (`MessageId`,
/// `Subject`, `BodyText`, `Unread`, `ReceivedUtc`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EmailMessage {
    pub message_id: String,
    pub subject: String,
    pub body_text: String,
    pub unread: bool,
    pub received_utc: DateTime<Utc>,
}

impl EmailMessage {
    /// Constructs an email message.
    pub fn new(
        message_id: impl Into<String>,
        subject: impl Into<String>,
        body_text: impl Into<String>,
        unread: bool,
        received_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            message_id: message_id.into(),
            subject: subject.into(),
            body_text: body_text.into(),
            unread,
            received_utc,
        }
    }
}

/// (Integration) An email connector.
///
/// Mirrors `interface IEmailConnector`.
pub trait IEmailConnector {
    fn provider_id(&self) -> String;
    fn is_configured(&self) -> bool;
    /// Unread messages, newest first, capped at `max` (`max < 0` yields nothing).
    fn list_unread(&self, max: i32) -> Vec<EmailMessage>;
    /// Messages whose subject or body contains `query` (case-insensitive), newest
    /// first, capped at `max`.
    fn search(&self, query: &str, max: i32) -> Vec<EmailMessage>;
    /// Marks a message read (no-op for an unknown id).
    fn mark_read(&self, message_id: &str);
}

/// (Integration) In-memory [`IEmailConnector`]: seeded with messages; unread +
/// search read newest-first, [`mark_read`](IEmailConnector::mark_read) flips the
/// flag.
///
/// Mirrors `sealed class InMemoryEmailConnector`.
pub struct InMemoryEmailConnector {
    messages: Mutex<HashMap<String, EmailMessage>>,
}

impl InMemoryEmailConnector {
    /// Creates an empty connector.
    pub fn new() -> Self {
        Self {
            messages: Mutex::new(HashMap::new()),
        }
    }

    /// Creates a connector seeded with `seed`.
    pub fn seeded(seed: impl IntoIterator<Item = EmailMessage>) -> Self {
        let mut map = HashMap::new();
        for m in seed {
            map.insert(m.message_id.clone(), m);
        }
        Self {
            messages: Mutex::new(map),
        }
    }
}

impl Default for InMemoryEmailConnector {
    fn default() -> Self {
        Self::new()
    }
}

impl IEmailConnector for InMemoryEmailConnector {
    fn provider_id(&self) -> String {
        "in-memory".to_string()
    }

    fn is_configured(&self) -> bool {
        true
    }

    fn list_unread(&self, max: i32) -> Vec<EmailMessage> {
        let take = max.max(0) as usize;
        let mut hits: Vec<EmailMessage> = self
            .messages
            .lock()
            .unwrap()
            .values()
            .filter(|m| m.unread)
            .cloned()
            .collect();
        hits.sort_by(|a, b| b.received_utc.cmp(&a.received_utc));
        hits.truncate(take);
        hits
    }

    fn search(&self, query: &str, max: i32) -> Vec<EmailMessage> {
        let take = max.max(0) as usize;
        let needle = query.to_lowercase();
        let mut hits: Vec<EmailMessage> = self
            .messages
            .lock()
            .unwrap()
            .values()
            .filter(|m| {
                m.subject.to_lowercase().contains(&needle)
                    || m.body_text.to_lowercase().contains(&needle)
            })
            .cloned()
            .collect();
        hits.sort_by(|a, b| b.received_utc.cmp(&a.received_utc));
        hits.truncate(take);
        hits
    }

    fn mark_read(&self, message_id: &str) {
        if let Some(m) = self.messages.lock().unwrap().get_mut(message_id) {
            m.unread = false;
        }
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// News
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) A news item.
///
/// Mirrors the `NewsItem` shape consumed by `INewsSource` (`ItemId`,
/// `PublishedUtc`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NewsItem {
    pub item_id: String,
    pub title: String,
    pub published_utc: DateTime<Utc>,
}

impl NewsItem {
    /// Constructs a news item.
    pub fn new(
        item_id: impl Into<String>,
        title: impl Into<String>,
        published_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            item_id: item_id.into(),
            title: title.into(),
            published_utc,
        }
    }
}

/// (Integration) A news source.
///
/// Mirrors `interface INewsSource`.
pub trait INewsSource {
    fn source_id(&self) -> String;
    fn is_configured(&self) -> bool;
    /// The latest items, newest first, capped at `max`.
    fn fetch_latest(&self, max: i32) -> Vec<NewsItem>;
}

/// (Integration) In-memory [`INewsSource`]: seeded items, newest-first.
///
/// Mirrors `sealed class InMemoryNewsSource`.
pub struct InMemoryNewsSource {
    items: Mutex<HashMap<String, NewsItem>>,
}

impl InMemoryNewsSource {
    /// Creates an empty source.
    pub fn new() -> Self {
        Self {
            items: Mutex::new(HashMap::new()),
        }
    }

    /// Creates a source seeded with `seed`.
    pub fn seeded(seed: impl IntoIterator<Item = NewsItem>) -> Self {
        let mut map = HashMap::new();
        for i in seed {
            map.insert(i.item_id.clone(), i);
        }
        Self {
            items: Mutex::new(map),
        }
    }
}

impl Default for InMemoryNewsSource {
    fn default() -> Self {
        Self::new()
    }
}

impl INewsSource for InMemoryNewsSource {
    fn source_id(&self) -> String {
        "in-memory".to_string()
    }

    fn is_configured(&self) -> bool {
        true
    }

    fn fetch_latest(&self, max: i32) -> Vec<NewsItem> {
        let take = max.max(0) as usize;
        let mut hits: Vec<NewsItem> = self.items.lock().unwrap().values().cloned().collect();
        hits.sort_by(|a, b| b.published_utc.cmp(&a.published_utc));
        hits.truncate(take);
        hits
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Weather
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) A weather sample.
///
/// Mirrors `sealed record WeatherSample(DateTimeOffset AtUtc, double TempC,
/// double FeelsLikeC, double PrecipMm, double WindKph, int HumidityPct,
/// string Condition)`.
#[derive(Debug, Clone, PartialEq)]
pub struct WeatherSample {
    pub at_utc: DateTime<Utc>,
    pub temp_c: f64,
    pub feels_like_c: f64,
    pub precip_mm: f64,
    pub wind_kph: f64,
    pub humidity_pct: i32,
    pub condition: String,
}

impl WeatherSample {
    /// Constructs a weather sample.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        at_utc: DateTime<Utc>,
        temp_c: f64,
        feels_like_c: f64,
        precip_mm: f64,
        wind_kph: f64,
        humidity_pct: i32,
        condition: impl Into<String>,
    ) -> Self {
        Self {
            at_utc,
            temp_c,
            feels_like_c,
            precip_mm,
            wind_kph,
            humidity_pct,
            condition: condition.into(),
        }
    }
}

/// (Integration) A weather provider.
///
/// Mirrors `interface IWeatherProvider`.
pub trait IWeatherProvider {
    fn provider_id(&self) -> String;
    /// Current conditions for a lat/lon.
    fn current(&self, lat: f64, lon: f64) -> WeatherSample;
    /// `hours` samples starting now for a lat/lon (`hours < 0` yields nothing).
    fn hourly(&self, lat: f64, lon: f64, hours: i32) -> Vec<WeatherSample>;
}

/// (Integration) In-memory [`IWeatherProvider`]: deterministic pseudo-weather
/// derived from coordinates + hour (no randomness, reproducible across platforms).
///
/// Mirrors `sealed class InMemoryWeatherProvider`.
pub struct InMemoryWeatherProvider;

impl InMemoryWeatherProvider {
    /// Creates the provider.
    pub fn new() -> Self {
        Self
    }

    /// The deterministic sample for a coordinate + hour offset. Mirrors the C#
    /// `Sample`: `tempC = round(15 + 10·cos((lat + hourOffset)·π/12), 2)`,
    /// `feelsLike = round(tempC − 1.5, 2)`, timestamp = Unix epoch + `hourOffset`.
    fn sample(lat: f64, _lon: f64, hour_offset: i32) -> WeatherSample {
        let temp_c = round_even(
            15.0 + 10.0 * ((lat + hour_offset as f64) * PI / 12.0).cos(),
            2,
        );
        let at_utc = Utc.timestamp_opt(0, 0).unwrap() + Duration::hours(hour_offset as i64);
        WeatherSample::new(
            at_utc,
            temp_c,
            round_even(temp_c - 1.5, 2),
            0.0,
            12.0,
            40,
            "Clear",
        )
    }
}

impl Default for InMemoryWeatherProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl IWeatherProvider for InMemoryWeatherProvider {
    fn provider_id(&self) -> String {
        "in-memory".to_string()
    }

    fn current(&self, lat: f64, lon: f64) -> WeatherSample {
        Self::sample(lat, lon, 0)
    }

    fn hourly(&self, lat: f64, lon: f64, hours: i32) -> Vec<WeatherSample> {
        (0..hours.max(0)).map(|h| Self::sample(lat, lon, h)).collect()
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Routing
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) A route estimate.
///
/// Mirrors `sealed record RouteEstimate(double DistanceKm, TimeSpan Duration,
/// IReadOnlyList<(double, double)> Polyline)`.
#[derive(Debug, Clone, PartialEq)]
pub struct RouteEstimate {
    pub distance_km: f64,
    pub duration: Duration,
    pub polyline: Vec<(f64, f64)>,
}

impl RouteEstimate {
    /// Constructs a route estimate.
    pub fn new(distance_km: f64, duration: Duration, polyline: Vec<(f64, f64)>) -> Self {
        Self {
            distance_km,
            duration,
            polyline,
        }
    }
}

/// (Integration) A routing provider.
///
/// Mirrors `interface IRoutingProvider`.
pub trait IRoutingProvider {
    fn provider_id(&self) -> String;
    /// A deterministic great-circle route estimate for `mode`
    /// (`"walk"`/`"bike"`/`"transit"`, else car).
    fn route(
        &self,
        from_lat: f64,
        from_lon: f64,
        to_lat: f64,
        to_lon: f64,
        mode: &str,
    ) -> RouteEstimate;
}

/// (Integration) In-memory [`IRoutingProvider`]: great-circle distance and a
/// mode-based speed give a deterministic estimate with a 2-point polyline.
///
/// Mirrors `sealed class InMemoryRoutingProvider`.
pub struct InMemoryRoutingProvider;

impl InMemoryRoutingProvider {
    /// Creates the provider.
    pub fn new() -> Self {
        Self
    }

    /// Great-circle distance (km) between two coordinates. Mirrors the C#
    /// `Haversine` exactly (`r = 6371`).
    pub fn haversine(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
        const R: f64 = 6371.0;
        let d_lat = (lat2 - lat1) * PI / 180.0;
        let d_lon = (lon2 - lon1) * PI / 180.0;
        let a = (d_lat / 2.0).sin() * (d_lat / 2.0).sin()
            + (lat1 * PI / 180.0).cos()
                * (lat2 * PI / 180.0).cos()
                * (d_lon / 2.0).sin()
                * (d_lon / 2.0).sin();
        R * 2.0 * a.sqrt().atan2((1.0 - a).sqrt())
    }
}

impl Default for InMemoryRoutingProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl IRoutingProvider for InMemoryRoutingProvider {
    fn provider_id(&self) -> String {
        "in-memory".to_string()
    }

    fn route(
        &self,
        from_lat: f64,
        from_lon: f64,
        to_lat: f64,
        to_lon: f64,
        mode: &str,
    ) -> RouteEstimate {
        let km = Self::haversine(from_lat, from_lon, to_lat, to_lon);
        let kph = match mode {
            "walk" => 5.0,
            "bike" => 18.0,
            "transit" => 30.0,
            _ => 60.0,
        };
        let hours = if kph <= 0.0 { 0.0 } else { km / kph };
        // TimeSpan.FromHours(hours) → whole milliseconds (chrono Duration is
        // integral); the C# TimeSpan carries sub-ms ticks, but the port keeps ms
        // resolution consistent with the rest of the crate.
        let duration = Duration::milliseconds((hours * 3_600_000.0).round() as i64);
        RouteEstimate::new(
            round_even(km, 3),
            duration,
            vec![(from_lat, from_lon), (to_lat, to_lon)],
        )
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Home automation
// ─────────────────────────────────────────────────────────────────────────────

/// (Integration) A home-automation entity.
///
/// Mirrors `sealed record HaEntity(string EntityId, string Domain, string State,
/// string FriendlyName)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HaEntity {
    pub entity_id: String,
    pub domain: String,
    pub state: String,
    pub friendly_name: String,
}

impl HaEntity {
    /// Constructs an entity.
    pub fn new(
        entity_id: impl Into<String>,
        domain: impl Into<String>,
        state: impl Into<String>,
        friendly_name: impl Into<String>,
    ) -> Self {
        Self {
            entity_id: entity_id.into(),
            domain: domain.into(),
            state: state.into(),
            friendly_name: friendly_name.into(),
        }
    }
}

/// (Integration) A home-automation connector.
///
/// Mirrors `interface IHomeAutomationConnector`.
pub trait IHomeAutomationConnector {
    fn provider_id(&self) -> String;
    fn is_configured(&self) -> bool;
    /// All entities, ordered by entity id.
    fn list_entities(&self) -> Vec<HaEntity>;
    /// Calls a service against every entity in `domain` (case-insensitive):
    /// `turn_on`→`on`, `turn_off`→`off`, `toggle` flips on/off, anything else is a
    /// no-op.
    fn call_service(&self, domain: &str, service: &str);
}

/// (Integration) In-memory [`IHomeAutomationConnector`]: seeded entities;
/// turn_on/turn_off/toggle deterministically mutate matching-domain entity state.
///
/// Mirrors `sealed class InMemoryHomeAutomationConnector`.
pub struct InMemoryHomeAutomationConnector {
    entities: Mutex<HashMap<String, HaEntity>>,
}

impl InMemoryHomeAutomationConnector {
    /// Creates an empty connector.
    pub fn new() -> Self {
        Self {
            entities: Mutex::new(HashMap::new()),
        }
    }

    /// Creates a connector seeded with `seed`.
    pub fn seeded(seed: impl IntoIterator<Item = HaEntity>) -> Self {
        let mut map = HashMap::new();
        for e in seed {
            map.insert(e.entity_id.clone(), e);
        }
        Self {
            entities: Mutex::new(map),
        }
    }
}

impl Default for InMemoryHomeAutomationConnector {
    fn default() -> Self {
        Self::new()
    }
}

impl IHomeAutomationConnector for InMemoryHomeAutomationConnector {
    fn provider_id(&self) -> String {
        "in-memory".to_string()
    }

    fn is_configured(&self) -> bool {
        true
    }

    fn list_entities(&self) -> Vec<HaEntity> {
        let mut hits: Vec<HaEntity> = self.entities.lock().unwrap().values().cloned().collect();
        hits.sort_by(|a, b| a.entity_id.cmp(&b.entity_id));
        hits
    }

    fn call_service(&self, domain: &str, service: &str) {
        let mut entities = self.entities.lock().unwrap();
        for e in entities.values_mut() {
            if !e.domain.eq_ignore_ascii_case(domain) {
                continue;
            }
            let new_state = match service {
                "turn_on" => "on".to_string(),
                "turn_off" => "off".to_string(),
                "toggle" => {
                    if e.state == "on" {
                        "off".to_string()
                    } else {
                        "on".to_string()
                    }
                }
                _ => e.state.clone(),
            };
            e.state = new_state;
        }
    }
}
